keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,886 @@
+      parse them first.
+
+    - (Only relevant to IPv6 because all of the non-canonical forms
+      for IPv4 addresses are now invalid.)
+
+    - (It doesn't make sense to force new values of existing API
+      fields to be canonical, because the "compare them as strings"
+      thing doesn't work if you have to worry about some objects
+      having legacy non-canonical values.)
+
+<<[/UNRESOLVED]>>
+
+
+<<[UNRESOLVED non-ip-fields ]>>
+
+- MAYBE tighten/make-consistent the validation of other
+  networking-related fields at the same time. In particular, some
+  fields that are supposed to be hostname-valued also accept IPv4 (but
+  not IPv6) addresses, because they just use
+  `utilvalidation.IsDNS1123Subdomain`, which requires basically
+  "alphanumeric separated by dots", which IPv4 addresses are.
+
+<<[/UNRESOLVED]>>
+
+
+<<[UNRESOLVED non-special-ip ]>>
+
+- MAYBE revisit the use of `ValidateNonSpecialIP`: certain kinds of
+  IPs (loopback, multicast, link-local, etc) do not make sense in many
+  contexts, but we only prohibit them in a few places.
+
+<<[/UNRESOLVED]>>
+
+
+<<[UNRESOLVED cli-validation ]>>
+
+- MAYBE change/consistentify the validation of IP/CIDR values in CLI
+  args / component configs. There is less room for changing the
+  mandatory validation here without badly breaking people, but we
+  could at least add warnings.
+
+<<[/UNRESOLVED]>>
+```
+
+[CEL IP/CIDR validation helpers]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/cel/library/ip.go
+
+### Non-Goals
+
+- ...
+
+## Proposal
+
+### Validation Changes
+
+#### Updated validation criteria
+
+IP and CIDR validation is currently handled by
+`netutils.ParseIPSloppy` and `netutils.ParseCIDRSloppy`, which are
+simply forks of the golang 1.17 versions of `net.ParseIP` and
+`net.ParseCIDR`, to preserve the historic semantics. These functions
+have various problems:
+
+- They allow IPv4 addresses to have leading "0"s: e.g.,
+  "`012.000.001.002`". The problem with this is that they parse these
+  by simply ignoring the extra "0"s (i.e., parsing "`012.000.001.002`"
+  as "`12.0.1.2`"), but most libc-based code treats the octets as
+  _octal_ in this case (i.e., parsing it as "`10.0.1.2`"). If such an
+  IP address is validated in golang code but then the original string
+  is passed to something libc-based, this may allow bypassing checks
+  on the IP range by taking advantage of the difference in parsing.
+  (This was [CVE-2021-29923].)
+
+- They allow "IPv4-mapped IPv6 addresses", e.g., "`::ffff:1.2.3.4`".
+  This form of IP address was part of an old IETF plan to simplify
+  dual-stack transition (by allowing an IPv4-only program to be
+  quickly converted to a syntactically-IPv6-only program that was
+  nonetheless semantically dual-stack capable). But they were never
+  widely used, and at this point mostly exist to cause confusion. This
+  confusion could, again, allow for smuggling IP addresses past
+  validation checks.
```

Comment:
it seems this kind of thing we need to live with it , see related discussion https://lwn.net/Articles/688462/

Reply:
I agree does not seems to have any use case and not having them should remove complexity ... if we can do it I agree we should remove them